### ISSUE TITLE
Support for Number (Integer/Float) and Null in serde_json

### ISSUE THREAD

**Issue by dramikei** [rusqlite/rusqlite#882]
I was trying to serialize query result into a serde_json Value inside `query_map` but I got an InvalidColumn error, which was a bit confusing because README.md claims that serde_json is supported (via cargo optional feature). 

After Looking around I found that the [serde_json.rs](https://github.com/rusqlite/rusqlite/blob/master/src/types/serde_json.rs) has FromSQL implementation only for `Text` and `Blob` types, and thus the error.

Is this by design or a mistake?

if a bug then I'd love to try and extend support for other serde_json types.

I am trying to implement a function in my application that would execute arbitrary SQLs and return a `json_serve::Map` vector using the column name as the key and the `json_serve::Value` as the Value.

**Comment by gwenn** [comment]
For [Number](https://docs.serde.rs/serde_json/value/enum.Value.html#variant.Number), I guess we should fix [FromSql](https://github.com/rusqlite/rusqlite/blob/master/src/types/serde_json.rs#L23) if SQLite type is [Integer](https://docs.rs/rusqlite/0.24.2/rusqlite/types/enum.Type.html#variant.Integer) or [Real](https://docs.rs/rusqlite/0.24.2/rusqlite/types/enum.Type.html#variant.Real).
For [Null](https://docs.serde.rs/serde_json/value/enum.Value.html#variant.Null), I am not sure: a "null" string is expected and not [NULL](https://docs.rs/rusqlite/0.24.2/rusqlite/types/enum.Type.html#variant.Null), no ?
For String, "\"text\"" is expected, right ?

For reference, [sqlx impl](https://docs.rs/sqlx-core/0.4.2/src/sqlx_core/sqlite/types/json.rs.html#33-44).

**Comment by gwenn** [comment]
```sh
sqlite> SELECT json_valid(NULL); -- not valid...
0
sqlite> SELECT json_valid('null');
1
sqlite> SELECT json_valid('text'); -- not valid...
0
sqlite> SELECT json_valid('"text"');
1
```
Should we [quote](https://sqlite.org/json1.html#jquote) ? When ?
```
sqlite> SELECT json_quote(NULL);
null
sqlite> SELECT json_quote('text');
"text"
sqlite>
```

**Comment by dramikei** [comment]
Not sure if I understood right, but I think we should just return serde_json's null type (`json!(null)` or maybe `serde_json::from_value(null)`) when we encounter ValueRef::Null.

**Comment by gwenn** [comment]
I mean SQL `NULL` is not a valid JSON value (only 'null' literal is valid).
And SQL literal 'text' is not a valid JSON value (only '"text"' is valid).

**Comment by dramikei** [comment]
Oh! I see, in case of Sql NULL cant we simply construct and return a
serde_json null?

in case of ‘test’ I guess wrapping it around “” is the only way.

**Comment by gwenn** [comment]
But for 'text', we don't know until parsed !

```patch
Index: src/types/serde_json.rs
===================================================================
diff --git a/src/types/serde_json.rs b/src/types/serde_json.rs
--- a/src/types/serde_json.rs	(revision d5b5f4301545d85d6cc08390ebb9a64930b5480e)
+++ b/src/types/serde_json.rs	(date 1611176061245)
@@ -1,6 +1,6 @@
 //! [`ToSql`] and [`FromSql`] implementation for JSON `Value`.
 
-use serde_json::Value;
+use serde_json::{Number,Value};
 
 use crate::types::{FromSql, FromSqlError, FromSqlResult, ToSql, ToSqlOutput, ValueRef};
 use crate::Result;
@@ -18,9 +18,16 @@
     #[inline]
     fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self> {
         match value {
-            ValueRef::Text(s) => serde_json::from_slice(s),
+            ValueRef::Text(s) => serde_json::from_slice(s), // KO for b"text"
             ValueRef::Blob(b) => serde_json::from_slice(b),
-            _ => return Err(FromSqlError::InvalidType),
+            ValueRef::Integer(i) => Ok(Value::Number(Number::from(i))),
+            ValueRef::Real(f) => {
+                match Number::from_f64(f) {
+                    Some(n) => Ok(Value::Number(n)),
+                    _ => return Err(FromSqlError::InvalidType), // FIXME
+                }
+            },
+            ValueRef::Null => Ok(Value::Null),
         }
         .map_err(|err| FromSqlError::Other(Box::new(err)))
     }
```

**Comment by dramikei** [comment]
You're right, SQL TEXT can have issues especially if it has quotes in it. I did a bit of testing and heres what I found:

// Let column at index 0 ==> Hello
// Let a Text value at row 0 ==> hello"world"

Heres a snippet of the function I've been writing (For Dynamic Sql -> JSON conversion)
```
let mut stmt = db.prepare(sql).unwrap();
  let rus = stmt.query_map(bindings, |row| {
    let mut query = String::from("{");
    for col_indx in 0 .. row.column_count() {
      
      match row.get(col_indx) {
        Ok(rusqlite::types::Value::Null) => query = query.clone() + format!("\"{}\":null,", row.column_name(col_indx).unwrap()).as_mut_str(),
        Ok(rusqlite::types::Value::Integer(x)) => query = query.clone() + format!("\"{}\":{:},", row.column_name(col_indx).unwrap(), x).as_mut_str(),
        Ok(rusqlite::types::Value::Real(x)) => query = query.clone() + format!("\"{}\":{:},", row.column_name(col_indx).unwrap(), x).as_mut_str(),

        // Format using {:?} escapes characters (but may escape incorrectly), using {} causes problems with quotes
        Ok(rusqlite::types::Value::Text(x)) => query = query.clone() + format!("\"{}\":{:?},", row.column_name(col_indx).unwrap(), x).as_mut_str(),
        Ok(rusqlite::types::Value::Blob(x)) => query = query.clone() + format!("\"{}\":{:?},", row.column_name(col_indx).unwrap(), x).as_mut_str(),
        _ => ()
      }
    }
    let _ = query.pop();         // Remove "," character
    query = query.clone() + "}"; // Append closing bracket
    Ok(query.clone())
  })?;
  let mut final_result: Vec<Value> = Vec::new();
  for result in rus {
    let v: Value = serde_json::from_str(&result.unwrap()).unwrap();
    final_result.push(v);
  }
  println!("{:?}", final_result);
  Ok(final_result)
```

// prints [Object({"Hello": String(hello“world”)})]

when I use {:?} in either println!() or format!() the quotes are escaped (But still this code had crashed a few times randomly so I believe that escaping isnt accurate), and they arent if I use {}. Maybe this can be of some use to us? Or we can look into the r#""# operator. But either way there are some edge cases still left.

**Comment by gwenn** [comment]
So it appears that your SQLite values are not really JSON values (serialized to text) !

**Comment by dramikei** [comment]
Yeah I was converting it into json using serde_json’s constructor. But the
point here is that I was able to overcome the raw string problem by using
{:?} in format!()

**Comment by gwenn** [comment]
I am reluctant to put this logic in `rusqlite`.

**Comment by dramikei** [comment]
Yeah not the best :P there would be a much much better way though

**Comment by thomcc** [comment]
Perhaps using https://crates.io/crates/serde-transcode and https://crates.io/crates/serde_rusqlite in conjunction will help you.

(Note: I've never used these crates, and can't vouch for their quality, I just know of their existence)

**Comment by xtuc** [comment]
A similar issue exists for the ToSQL implementation: https://github.com/rusqlite/rusqlite/blob/master/src/types/serde_json.rs#L12. All JSON values end up as SQL strings. I believe this is straightforward to fix, would be fine if I sent a patch?

**Comment by gwenn** [comment]
> A similar issue exists for the ToSQL implementation: https://github.com/rusqlite/rusqlite/blob/master/src/types/serde_json.rs#L12. All JSON values end up as SQL strings. I believe this is straightforward to fix, would be fine if I sent a patch?

For reference, [sqlx_impl](https://github.com/launchbadge/sqlx/blob/main/sqlx-core/src/sqlite/types/json.rs#L27).
